titus-server-master/src/main/java/com/netflix/titus/master/scheduler/resourcecache/TaskCache.java

```diff
@@ -116,28 +134,69 @@ private void buildTaskCacheInfo(List<Pair<Job, List<Task>>> jobsAndTasks) {
                         getIpAllocationZone(ipAllocationId, jobAndTask.getLeft().getJobDescriptor()).ifPresent(zoneIdForIpAllocation ->
                                 ipAllocationIdToZoneId.put(ipAllocationId, zoneIdForIpAllocation));
                     });
+
+                    getOpportunisticCpuAllocation(task).ifPresent(allocation -> {
+                        if (TaskState.isRunning(task.getStatus().getState())) {
+                            CollectionsExt.multiMapAddValue(assignedOpportunisticCpus, allocation.getAgentId(), allocation);
+                        }
+                    });
                 }
                 zoneBalanceCountersByJobId.put(jobAndTask.getLeft().getId(), jobZoneBalancing);
             }
         }
 
-        private String getZoneId(Task task) {
-            return task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_ZONE);
-        }
-
-        private Optional<String> getIpAllocationId(Task task) {
-            return Optional.ofNullable(task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_IP_ALLOCATION_ID));
-        }
-
         private Optional<String> getIpAllocationZone(String ipAllocationId, JobDescriptor<?> jobDescriptor) {
             for (SignedIpAddressAllocation signedIpAddressAllocation : jobDescriptor.getContainer().getContainerResources().getSignedIpAddressAllocations()) {
                 if (signedIpAddressAllocation.getIpAddressAllocation().getAllocationId().equals(ipAllocationId)) {
                     return Optional.of(signedIpAddressAllocation.getIpAddressAllocation().getIpAddressLocation().getAvailabilityZone());
                 }
             }
-            titusRuntime.getCodeInvariants().inconsistent("Unable to find zone for IP allocation ID {} in job allocations {}",
+            codeInvariants().inconsistent("Unable to find zone for IP allocation ID {} in job allocations {}",
                     ipAllocationId, jobDescriptor.getContainer().getContainerResources().getSignedIpAddressAllocations());
             return Optional.empty();
         }
+
+        private Optional<OpportunisticCpuAllocation> getOpportunisticCpuAllocation(Task task) {
+            Optional<String> allocationIdOpt = getOpportunisticCpuAllocationId(task);
+            Optional<OpportunisticCpuAllocation> allocationOpt = allocationIdOpt
```

Comment:
Optimize?

Reply:
did you mean `inline`? I'm using the `allocationIdOpt` variable again below

Reply:
No, it is again about not using functional API, as it will produce several temporal objects.